File: 240929/안전 지대/comfort-zone.py
#k를 초과하면 뭉텅이 취급
#뭉텅이의 개수가 최대인 경우를 찾아라
#최대일 때 뭉텅이 개수, 그때의 K (여러 개면 최솟값)

n, m = map(int, input().split())
area = []
for i in range(n):
    area.append(list(map(int, input().split())))

max_k = 0
for row in area:
    for col in row:
        max_k = max(max_k, col)
min_k = max_k
for row in area:
    for col in row:
        min_k = min(min_k, col)

def in_range(x, y):
    if 0 <= x < n and 0 <= y < m:
        return True
    return False

def can_go(x, y):
    if in_range(x, y) and area[x][y] > k and not visited[x][y]:
        return True
    return False

def dfs(x, y):
    dx, dy = [0, 1, 0, -1], [1, 0, -1, 0]

    for dr in range(4):
        nx, ny = x + dx[dr], y + dy[dr]

        if can_go(nx, ny):
            visited[nx][ny] = 1
            dfs(nx, ny)
    
klst = []
for k in range(min_k, max_k + 1):
    visited = []
    for i in range(n):
        visited.append([0] * m)

    comfort = 0
    for row in range(n):
        for col in range(m):
            if can_go(row, col):
                visited[row][col] = 1
                comfort += 1
                dfs(row, col)
    
    klst.append([k, comfort])

klst.sort(key = lambda x: (-x[1], x[0]))
print(klst[0][0], klst[0][1])

File: 250222/안전 지대/comfort-zone.cpp
#include <iostream>
using namespace std;

int n, m;
int area[50][50];
bool visited[50][50];
int dx[] = { -1, 0, 1, 0 };
int dy[] = { 0, 1, 0, -1 };

// k이하는 다 가라앉음
// 안전 영역의 수가 최대가 될 때의 k, 그때의 안전 영역

bool InRange(int x, int y) {
    return (0 <= x && x < n && 0 <= y && y < m);
}

bool CanGo(int x, int y, int k) {
    return InRange(x, y) && area[x][y] > k && !visited[x][y];
}

void DFS(int x, int y, int k) {
    for (int dr = 0; dr < 4; dr++) {
        int nx = x + dx[dr];
        int ny = y + dy[dr];

        if (CanGo(nx, ny, k)) {
            visited[nx][ny] = true;
            DFS(nx, ny, k);
        }
    }
}

int main() {
    cin >> n >> m;
    for (int i = 0; i < n; i++)
        for (int j = 0; j < m; j++)
            cin >> area[i][j];

    int min_k = 100, max_k = 0;
    for (int i = 0; i < n; i++)
        for (int j = 0; j < m; j++) {
            min_k = min(min_k, area[i][j]);
            max_k = max(max_k, area[i][j]);
        }

    int ans = 1;
    int safe_zone = 0;

    for (int k = max(1, min_k-1); k <= max_k; k++) {
        int cnt = 0;
        for (int i = 0; i < n; i++)
            for (int j = 0; j < m; j++)
                visited[i][j] = false;

        for (int i = 0; i < n; i++)
            for (int j = 0; j < m; j++) {
                if (area[i][j] > k && !visited[i][j]) {
                    visited[i][j] = true;
                    cnt++;
                    DFS(i, j, k);
                }
            }

        if (cnt > safe_zone) {
            safe_zone = cnt;
            ans = k;
        }
    }

    cout << ans << " " << safe_zone;
    return 0;
}